src/bos/operators/utils/clients/pcs.py

```diff
@@ -202,6 +209,10 @@ def power_on(nodes, session=None, task_deadline_minutes=1, **kwargs):
     Sends a request to PCS for transitioning nodes in question to a powered on state.
     Returns: A JSON parsed object response from PCS, which includes the created request ID.
     """
+    if not nodes:
+        # Should probably raise an exception here, since we don't want to actually call PCS
+        # with an empty node list. Suggestions welcome for what to raise.
```

Comment:
Here are a list of the exceptions I found.

\]jasonsollom@Jasons-MacBook-Pro bos % \[\]grep -RE "^ *class [A-Z]" *|grep -Ei "Exception"
common/tenant_utils.py:class InvalidTenantException(Exception):
operators/session_setup.py:class SessionSetupException(Exception):
operators/utils/clients/hsm.py:class HWStateManagerException(Exception):
operators/utils/clients/s3.py:class ArtifactNotFound(Exception):
operators/utils/clients/s3.py:class ManifestNotFound(Exception):
operators/utils/clients/s3.py:class TooManyArtifacts(Exception):
operators/utils/clients/s3.py:class S3MissingConfiguration(Exception):
operators/utils/clients/s3.py:class S3ObjectNotFound(Exception):
operators/utils/clients/pcs.py:class PowerControlException(Exception):
operators/utils/clients/pcs.py:class PowerControlSyntaxException(Exception):
operators/utils/clients/pcs.py:class PowerControlTimeoutException(PowerControlException):
operators/utils/rootfs/__init__.py:class ProviderNotImplemented(Exception):
operators/utils/__init__.py:class ServiceNotReady(Exception):
operators/utils/boot_image_metadata/__init__.py:class BootImageMetaDataBadRead(Exception):
operators/utils/boot_image_metadata/factory.py:class BootImageMetaDataUnknown(Exception):
operators/base.py:class BaseOperatorException(Exception):
operators/base.py:class MissingSessionData(BaseOperatorException):
reporter/node_identity.py:class UnknownIdentity(BOSException):
reporter/__init__.py:class BOSException(Exception):
reporter/components/__init__.py:class BOSComponentException(BOSException):
reporter/components/state.py:class UnknownComponent(BOSComponentException):
reporter/components/state.py:class UnrecognizedResponse(BOSComponentException):
server/dbs/boot_artifacts.py:class BssTokenException(Exception):
server/dbs/boot_artifacts.py:class BssTokenUnknown(BssTokenException):
server/controllers/v2/sessions.debug.py:class ParsingException(Exception):
server/controllers/v2/sessions.py:class ParsingException(Exception):

None really fits the bill.  The closest is 
operators/utils/__init__.py:class ServiceNotReady(Exception):

You could create an addition exception in here called ComponentsEmpty(Exception) and use it when the PCS calls experience empty components.

Reply:
Done.